### PR TITLE
Command-tabbing fix

### PR DESCRIPTION
**Pull request by IntellectualKitty**
The new code uses presentation options to properly configure the application for fullscreen operation when appropriate.

Also, the window id is supplied as a parameter to the orderFront, makeKeyAndOrderFront, and orderOut calls instead of nil, which I *believe* is more correct.

### CODE REVIEW

**Comment by IntellectualKitty** [comment]
Closed due to lack of activity.

**Comment by Shimmen** [comment]
I don't see how closing the pull request will help achieve anything. The owners of GLFW are busy and as long as it's an open pull request it's on a list of things to consider pulling in. The code you've added seem good to me so just keep it open and wait.

**Comment by elmindreda** [comment]
What @Shimmen said.

**Comment by IntellectualKitty** [comment]
@Shimmen I appreciate your thoughts.

While I really like the clean and easy to use interface that GLFW provides, at its current level of development it's not suitable for my projects.  So, in my work, I use SDL (www.libsdl.org).

SDL provides a superset of the functionality of GLFW.  In fact, you could easily implement GLFW as a thin layer on top of SDL and (1) you'd get lots of additional functionality such as joystick events, proper hat switch and joystick trackball support, game controller support, touch device and gesture support, message boxes, etc. and (2) it would work on more platforms including iOS and Android.

With the PRs that I've submitted to GLFW, I've tried to address some of these deficiencies such as hat switch support #906 and joystick axis and button event support #856.  However, even at a more complete level of development, GLFW still would not provide the same level of cross-platform support that SDL does.  So, I never expected to benefit from these changes personally.

So, my real intention in making these PRs was to try to contribute, to try to do something useful that would help others.  However, when weeks and months go by and my PRs receive no activity, they serve no purpose except as persistent reminders that my time and energy has been spent fruitlessly.

So, while I appreciate your perspective, there's really no benefit for me not to close inactive PRs.

**Comment by elmindreda** [comment]
GLFW isn't a replacement for SDL2 and has never tried to be. I also can't in my spare time keep up with a project that has paid full-time developers and the backing of Valve and Epic. SDL2 is an excellent choice and I'm certainly not going to argue against it.

Thank you for your PRs; I'll get to them as soon as I can. Some of them are significant API additions and so require uninterrupted time for review in a way that I haven't had lately.

**Comment by IntellectualKitty** [comment]
I love GLFW and its clean interface, and I would gladly use it exclusively if it did what I needed.  For instance, I can easily live without the PRs I submitted for joystick button and axis events and for hat switch support, but I need the fullscreen functionality on macOS that I submitted with this PR.  So, please don't misunderstand my comments comparing SDL and GLFW.  Again, I prefer GLFW but I use SDL because I have to.

You've done a great job shepherding GLFW and I understand how important that role is for the quality of GLFW.  Of course your time is limited for such a demanding task.  And, of course, there's no way that one person could compete against the resources that SDL has.  However, I don't think that you're taking full advantage of the power of the open source community. 

There are people who are willing to help you.  That's the fundamental idea behind open source, namely that we can do more together than we can individually.  It's important for code quality to maintain strict control, but it also can be tremendously beneficial to use the resources that are available to you with the open source community.  For instance, you might consider taking on one or more partners to help you maintain GLFW.  Another idea might be to host one or more experimental branches that contain PRs with only preliminary approval and use feedback from the community to evaluate their effectiveness and quality.  I believe that the code that I submitted with this PR is correct, but it hasn't been tested on any other Macs; also, as you say, the code that I submitted with the other PRs made significant API changes.  If PRs like these were included into one or more experimental branches, you might be able to use the feedback regarding them to make evaluations -- rather than having to laboriously evaluate everything by yourself -- before incorporating them into releases or into the main branch.

These are just a couple of ideas off the top of my head.  And, of course, I am *not* yet an expert with GitHub, so they may not be as feasible as I think.  Nevertheless, I firmly believe in the value of the open source community and the idea that we can do more when we work together.  So, I would encourage you to try take more advantage of the help that is available to you.

**Comment by IntellectualKitty** [comment]
I've thought of another way that would make it easier to let people submit code, namely by bracketing their code changes with #ifdef / #endif directives.  For example, I could have easily bracketed the changes with this PR using #ifdef GLFW_COMMAND_TABBING_FIX so that the new code would be enabled only if GLFW_COMMAND_TABBING_FIX is defined in the main header file.  The code that I included with this PR was purely additional, so for changes that modify existing code you could have them keep the original code after an #else.  In the main header file, the #define statements for the macros could be commented out (with additional comments referring to the original PR) by default so that they are enabled only if people wish to use or test the new functionality leaving the original release functionality intact.

Using this type of approach, it should generally be relatively easy to confirm that changes are properly bracketed by #ifdef / #else / #endif directives when reviewing changed files on GitHub.  And when the time comes to integrate, modify, or remove the changed code, it should also be easy to accomplish with a global search.

Using this type of approach would make it easy to let people help and reduce your burden since your review would consist of a general understanding of the changes -- rather than having to make such a time-intensive review like you described above -- with a specific focus on confirming that new code is correctly bracketed by directives before you approve it.  You could clearly define the standards you want in documentation, and refuse to accept PRs until they meet those standards.  Then, like I mentioned in my previous comment, you can use feedback from people who do enable such changes to evaluate their merit and effectiveness.

I think this idea is better than the two ideas I mentioned previously since it should be very easy to accomplish this simply by (1) adding preliminary documentation regarding coding standards and how to properly bracket code changes and (2) taking advantage of the simplicity of GitHub's changed file viewer.

**Comment by elmindreda** [comment]
(Slip of the Comment button, part two follows)

There's often no way for a contributor to predict when a PR will save time without asking, so if you haven't already written a working fix, that's what I recommend you do. Open an issue and describe the problem, like you would for a PR. Of course, if the code is already written for your own immediate needs, make a PR of it.

Even PRs that don't save time are invaluable measured in [egoboo](https://en.wikipedia.org/wiki/egoboo), because someone cared enough about the project to take the time to develop and submit it, and I try to give egoboo back by keeping the original commit (possibly squashed) and adding credits even if large portions need rewriting. Others provide new knowledge, insights or even coding style changes that then spread throughout the project long after the submitter moved on to other ones. There are many examples of this.

So in summary, suggestions and feedback on existing and future features is great, both before and after they're merged. Helping people with common problems here or on the forum is great. Writing, discussing, testing and reviewing PRs is great. There is no ordering here. I'd put them all in first place if I could.

**Comment by elmindreda** [comment]
Then there's the elephant-in-the-room of me being pretty useless at interacting with people and giving them the recognition they deserve, which I know has cost the project both previous and potential contributors. Again, sorry about that. I'm working on it and may one day hand over the reins for that reason.

**Comment by IntellectualKitty** [comment]
@elmindreda, I want to apologize.  Clearly I have no idea what goes into maintaining a project like GLFW.  I thought about this only from my own perspective, namely that I tried to write the highest quality code that I could, and where I made API changes, that they matched existing GLFW standards, both for the code -- including coding style, function names and signatures, variables names, behavior, etc. -- and for the documentation comments.  My goal was to make my changes indistinguishable from original GLFW code and documentation.  So, without understanding the context of what actually goes into maintaining GLFW, I thought that what I had written would make a real contribution, and when it didn't go anywhere, I felt like I had wasted my time.

There is also the issue that I would really prefer to use GLFW rather than SDL, and when my PRs didn't go anywhere, particularly this one, it meant that I would have to maintain the additional code that I needed and manually re-integrate it every time I wanted to re-update to the master branch so that I could get the latest enhancements and bug fixes.

Another problem that I had was that every time I wanted to submit a new PR, I had to delete my existing repository and re-fork GLFW.  I just recently started to get a feel for branching (for a PR that I submitted to another project), but I am still not terribly proficient with GitHub.  So, if there is a way that I could have written new PRs against the master branch without having to re-fork -- or to merge GLFW's master branch back into a branch in my own fork so that I could keep up with changes that way -- I haven't yet found it.

I kept copies of all of the changed code on my hard drive.  So, if you can't merge these PRs into GLFW if and when you choose since I closed them (I can see that the commits and file changes are still here even though the original repositories have been deleted), please let me know and I will start over and resubmit them.

As for the experimental branches, I had no idea that you were already doing this.  The place that I would have looked for them is in the README on the front page (I know there are many branches under the branching tab, but I hadn't looked there because I thought they were just from other forks of GLFW rather than officially maintained branches, but I just took a look and found out that all 17 are yours).  Perhaps they would still get more traction if you had a section on actively developed branches in the README with the branch name and a brief description of its purpose along with a request for feedback.

Thank you for all the hard work that you do maintaining GLFW.  And, again, I apologize for my complete lack of understanding about how much work goes into it.

**Comment by elmindreda** [comment]
> I want to apologize.

There's no need to apologize and it was unfair of me to rant at you.

> I tried to write the highest quality code that I could, and where I made API changes, that they matched existing GLFW standards, both for the code -- including coding style, function names and signatures, variables names, behavior, etc. -- and for the documentation comments. My goal was to make my changes indistinguishable from original GLFW code and documentation.

From what I can tell looking through them now, you succeeded. My rant wasn't directed at your pending code, which I haven't yet reviewed. I've also already merged two of your PRs, earlier this year, and both of them have been included in releases without causing complaints, so that shouldn't worry you.

> So, without understanding the context of what actually goes into maintaining GLFW, I thought that what I had written would make a real contribution, and when it didn't go anywhere, I felt like I had wasted my time.

That is my fault and I apologize, to you and everyone else I've kept waiting. As a consequence of this thread I'll try setting aside some time (hopefully) each week to give initial feedback on at least one outstanding PR. That still won't be an instant response, but it shouldn't be months either.

> [...] when my PRs didn't go anywhere, particularly this one, it meant that I would have to maintain the additional code that I needed and manually re-integrate it every time I wanted to re-update to the master branch so that I could get the latest enhancements and bug fixes.
> Another problem that I had was that every time I wanted to submit a new PR, I had to delete my existing repository and re-fork GLFW. I just recently started to get a feel for branching (for a PR that I submitted to another project), but I am still not terribly proficient with GitHub. So, if there is a way that I could have written new PRs against the master branch without having to re-fork -- or to merge GLFW's master branch back into a branch in my own fork so that I could keep up with changes that way -- I haven't yet found it.

I don't know exactly what you know, so I'll probably explain things you know and leave out details you don't, but here goes...

One shouldn't ever have to re-fork. After you've forked this repo (or if you have a fork again, just keep it, it's fine and almost infinitely fixable), put all the changes for a given PR in a separate, non-master branch. Then open a PR from that branch. You may already have been doing that bit; I can't tell as GitHub lost that information when you deleted your fork. If your branch contains many commits that are really parts of the same change, squash them to a single commit before making the PR. It'll make it easier for everyone to work with.

Any changes you make to that branch (and only that branch) from then on will automatically update the PR, but if you don't need to update it for your own needs you can just leave it alone until I get to merging it. Once a PR has been reviewed and accepted, I'll start maintaining and periodically rebasing an experimental branch while it's being prepared for merging (current examples include `preedit-ime` and `egldevice`), but I won't do that for not-yet-accepted ones.

As for getting new changes, add a remote for this repo (let's call it `upstream` for now) to the local copy of your fork and fetch from it when you want new stuff. Then either rebase, merge or reset-and-merge your master branch (`origin/master`), which I'm guessing you want to contain the merged changes from one or more of your PRs, using the remote master (`upstream/master`). There are many possible models here.

If we switch places for a moment, I'd probably rebase each PR branch onto `upstream/master`, reset `origin/master` to `upstream/master` merge each refreshed PR branch into `origin/master` and then use that for my daily work until the PRs are merged. If the PR changes are isolated and uncomplicated like this one, I could probably skip the rebasing. That's just an example off the top of my head and there are probably better ways. It's been a while since I had to maintain simultaneous PRs for a single project.

The GitHub help has plenty of hands-on sections on every part of this and there are tons of tutorials out there that cater to different learning preferences.

> I kept copies of all of the changed code on my hard drive. So, if you can't merge these PRs into GLFW if and when you choose since I closed them (I can see that the commits and file changes are still here even though the original repositories have been deleted), please let me know and I will start over and resubmit them.

You don't even need to do that. I have a line in my local `.git/config` that fetches and caches all PRs as remote branches. This used to be easier to find in the GitHub help, but for anyone curious, the line is:
```
fetch = +refs/pull/*/head:refs/remotes/origin/pr/*
```
This lets me work with any existing PR even when offline, or in cases like this when the source disappears.

> As for the experimental branches, I had no idea that you were already doing this. The place that I would have looked for them is in the README on the front page (I know there are many branches under the branching tab, but I hadn't looked there because I thought they were just from other forks of GLFW rather than officially maintained branches

Branches created in forks won't ever appear in the branch list of this repo, even though you get all of the branches of this repo when you fork. Even when you make PRs of your own branches, they can only target branches that I or another maintainer has already created here.

> Perhaps they would still get more traction if you had a section on actively developed branches in the README with the branch name and a brief description of its purpose along with a request for feedback.

That's useful feedback and definitely worth a try. Thank you. I don't think I want to maintain an actual list in the README (too much update noise and risk of stale information), but I'll mention their existence and describe some of the work flow around them. If anyone would like to maintain a branch list with descriptions in the wiki, let me know and I'll arrange write permission.

**Comment by shurcooL** [comment]
> This used to be easier to find in the GitHub help

Just checking, but is [this](https://help.github.com/articles/checking-out-pull-requests-locally/) it?

> I am still not terribly proficient with GitHub

@IntellectualKitty, if you haven't already read it, I extremely highly recommend this article about how [git can be seen as a purely functional data structure](https://blog.jayway.com/2013/03/03/git-is-a-purely-functional-data-structure/); I think it helps a ton with becoming comfortable doing advanced git-fu (and using GitHub is a natural extension of that).

**Comment by elmindreda** [comment]
@shurcooL Yup, that's the page.

**Comment by IntellectualKitty** [comment]
@elmindreda and @shurcooL Thanks for all of your help.  I only just recently started branching like @elmindreda suggested, well after the PRs I submitted to GLFW however.  I'm afraid that much of @elmindreda 's comments are still a bit over my head (although the article on git as a purely functional data structure helped).

The main problem I've been having is that I've been using the web interface exclusively, thinking that it did everything.  I always try to be careful to avoid downloading malicious software, and I thought that the "GitHub Desktop" that shows up when I hit the "Clone or download" button was a toolbar that would install in my browser.  However, when I started researching how to get new changes back into my PRs after I last wrote, I finally found out that `git` is a command-line tool that does immensely more than the web interface.

@elmindreda I'll probably have to go back over your comments a few more times, but now that I know that there is a command-line interface, I think I'll be able to get up to speed fairly quickly.
